File: manual/data/java/tests/test_sklearn_OneHotEncoder.java
package com.example;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

public class test_sklearn_OneHotEncoder {

    private function_sklearn_OneHotEncoder encoder = new function_sklearn_OneHotEncoder();

    @Test
    public void test_function_call() {
        String data = "category1 5";
        try {
            String result = encoder.test_one_hot_encoding(data);
            assertTrue(result instanceof String);
        } catch (Exception e) {
            fail("Function call failed with exception: " + e);
        }
    }

    @Test
    public void test_correct_data() {
        String result = encoder.test_one_hot_encoding("category1 5");
        assertTrue(result instanceof String);
    }

    @Test
    public void test_encoded_data() {
        String result = encoder.test_one_hot_encoding("category1 5");
        String[] resultSplit = result.split(",");
        assertEquals("5", resultSplit[0]);
        assertEquals("1.0", resultSplit[1]);
    }

    @Test
    public void test_invalid_data_type() {
        // 在 Java 中由于静态类型限制，无法传入非 String，因此模拟传入 null
        Exception exception = assertThrows(RuntimeException.class, () -> {
            encoder.test_one_hot_encoding(null);
        });
        assertEquals("Input data must be a string.", exception.getMessage());
    }

    @Test
    public void test_missing_categorical_data() {
        Exception exception = assertThrows(RuntimeException.class, () -> {
            encoder.test_one_hot_encoding("12");
        });
        assertEquals("The input data has incorrect structure.", exception.getMessage());
    }
}

File: manual/data/cpp/tests/test_sklearn_OneHotEncoder.cpp
#include <gtest/gtest.h>
#include "../src/function_sklearn_OneHotEncoder.cpp"

class TestOneHotEncodingFunction : public ::testing::Test {
protected:
    OneHotEncoderFunction encoder;
};

TEST_F(TestOneHotEncodingFunction, TestFunctionCall) {
    std::string data = "category1 5";
    try {
        std::string result = encoder.test_one_hot_encoding(data);
        EXPECT_EQ(typeid(result).name(), typeid(std::string).name());
    } catch (const std::exception& e) {
        FAIL() << "Function call failed with exception: " << e.what();
    }
}

TEST_F(TestOneHotEncodingFunction, TestCorrectData) {
    std::string result = encoder.test_one_hot_encoding("category1 5");
    EXPECT_EQ(typeid(result).name(), typeid(std::string).name());
}

TEST_F(TestOneHotEncodingFunction, TestEncodedData) {
    std::string result = encoder.test_one_hot_encoding("category1 5");
    std::istringstream iss(result);
    std::string numeric_part, encoded_part;
    std::getline(iss, numeric_part, ',');
    std::getline(iss, encoded_part, ',');
    EXPECT_EQ(numeric_part, "5");
    EXPECT_EQ(encoded_part, "1");
}

TEST_F(TestOneHotEncodingFunction, TestInvalidDataType) {
    EXPECT_THROW(encoder.test_one_hot_encoding("12345"), std::invalid_argument);
}

TEST_F(TestOneHotEncodingFunction, TestMissingCategoricalData) {
    EXPECT_THROW(encoder.test_one_hot_encoding("12"), std::invalid_argument);
}

int main(int argc, char **argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}